data/src/main/java/io/gnosis/data/utils/SafeTxHash.kt

```diff
@@ -0,0 +1,108 @@
+package io.gnosis.data.utils
+
+import io.gnosis.data.models.DetailedExecutionInfo
+import io.gnosis.data.models.TransactionDetails
+import io.gnosis.data.models.TransactionInfo
+import io.gnosis.data.models.TransferInfo
+import pm.gnosis.crypto.utils.Sha3Utils
+import pm.gnosis.model.Solidity
+import pm.gnosis.utils.hexToByteArray
+import pm.gnosis.utils.toHex
+import java.math.BigInteger
+
+private const val ERC191_BYTE = "19"
+private const val ERC191_VERSION = "01"
+
+fun calculateSafeTxHash(
+    safeAddress: Solidity.Address, transaction: TransactionDetails
+): ByteArray? {
+
+    val to = when (val txInfo = transaction.txInfo) {
+        is TransactionInfo.Transfer -> {
+            when (val transferInfo = txInfo.transferInfo) {
+                is TransferInfo.Erc20Transfer -> {
+                    transferInfo.tokenAddress
+                }
+                is TransferInfo.Erc721Transfer -> {
+                    transferInfo.tokenAddress
+                }
+                is TransferInfo.EtherTransfer -> {
+                    txInfo.recipient
+                }
+            }
+        }
+        is TransactionInfo.Custom -> {
+            txInfo.to
+        }
+        is TransactionInfo.SettingsChange -> {
+            safeAddress
+        }
+        is TransactionInfo.Creation -> {
+            txInfo.factory
+        }
+        else -> {
+            throw UnsupportedTransactionType(transaction::javaClass.name)
+        }
+    }?.value.paddedHexString()
+
+    val value = transaction.txData?.value.paddedHexString()
+    val data = Sha3Utils.keccak(transaction.txData?.hexData?.hexToByteArray() ?: ByteArray(0)).toHex().padStart(64, '0')
+    val operationString = (transaction.txData?.operation?.id?.toBigInteger() ?: BigInteger.ZERO).paddedHexString()
+
+    val executionInfo = when (val executionInfo = transaction.detailedExecutionInfo) {
+        is DetailedExecutionInfo.MultisigExecutionDetails -> {
+            executionInfo
+        }
+        else -> {
+            throw UnsupportedTransactionType(transaction::javaClass.name)
+        }
```

Comment:
This is just a hard cast with extra steps 😄  I would require the `MultisigExecutionDetails` as an input to the method, so you don't need to hard cast everywhere.

Reply:
Why not use a hard cast here? It would have the same effect and `ClassCastException` is sufficiently informative?

Reply:
it's been updated. Please recheck